[ews-build.webkit.org] Check ownership of target branch

JonWBedard opened the pull request:
#### 038c4c107f86fd221227b3112c8c93435c37c3ce
<pre>
[ews-build.webkit.org] Check ownership of target branch
<a href="https://bugs.webkit.org/show_bug.cgi?id=242873">https://bugs.webkit.org/show_bug.cgi?id=242873</a>
&lt;rdar://problem/97225367&gt;

Reviewed by Aakash Jain.

* Tools/CISupport/ews-build/factories.py:
(MergeQueueFactoryBase.__init__): Add ValidateRemote.
* Tools/CISupport/ews-build/factories_unittest.py:
(TestExpectedBuildSteps):
* Tools/CISupport/ews-build/steps.py:
(ValidateRemote.__init__): Set default summary.
(ValidateRemote.start): Check if the branch as owned by the target remote exists
on the branch as owned by origin.
(ValidateRemote.getResultSummary): Set the step summary.
(ValidateRemote.evaluateCommand): Invert return code. If the branch as owned by the
target remote exists on origin, fail the build and block the PR.
(ValidateRemote.doStepIf): Skip step for patches and origin-targeting PRs.
(ValidateRemote.hideStepIf): Hide step if it&apos;s being skipped.
* Tools/CISupport/ews-build/steps_unittest.py:

Canonical link: <a href="https://commits.webkit.org/252584@main">https://commits.webkit.org/252584@main</a>
</pre>

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -4570,7 +4570,8 @@ def __init__(self, **kwargs):
         head_ref = self.getProperty('github.base.ref', 'main')
-        self.command = ['git', 'push', 'origin', f'HEAD:{head_ref}']  # FIXME: Support secret remotes
+        remote = self.getProperty('remote', '?')
+        self.command = ['git', 'push', remote, f'HEAD:{head_ref}']  # FIXME: Support secret remotes
```
what will happen to this command execution when remote is "?"

JonWBedard replied:
The `git push` command will fail because the remote doesn't exist.

aj062 commented on Tools/CISupport/ews-build/steps_unittest.py:
```diff
@@ -5738,6 +5743,60 @@ def test_failure(self):
+                command=['git', 'merge-base', '--is-ancestor', 'remotes/security/main', 'remotes/origin/main'],
+            ) + 0,
+        )
+        self.expectOutcome(result=FAILURE, state_string="Cannot land on 'main', it is owned by 'origin'")
```
This error message seems very confusing. What does it mean to be owned by 'origin'. Whom should it be owned by?

JonWBedard replied:
Maybe `WebKit/WebKit` is more accurate?

The branch should be owned by the remote which the PR is made against. So if I make a PR against `WebKit/WebKit-security`, it needs to be going to a branch owned by `WebKit/Webkit-security`.

aj062 replied:
Yeah, `WebKit/WebKit` might be better. Also, would be better to include action which user can take. For example something like:

"Cannot land on 'main' since this branch is owned by WebKit/Webkit repo, should create PR agains XYZ"

JonWBedard replied:
I'll add that to the PR comment.

aj062 commented on Tools/CISupport/ews-build/steps.py:
```diff
@@ -4769,6 +4770,61 @@ def start(self):
+        rc = super(ValidateRemote, self).evaluateCommand(cmd)
+
+        if rc == SUCCESS:
+            self.summary = f"Cannot land on '{base_ref}', it is owned by 'origin'"
```
This error message seems very confusing. What does it mean to be owned by 'origin'. Whom should it be owned by? Should be more explanative.

JonWBedard committed:
[ews-build.webkit.org] Check ownership of target branch
https://bugs.webkit.org/show_bug.cgi?id=242873
<rdar://problem/97225367>

Reviewed by Aakash Jain.

* Tools/CISupport/ews-build/factories.py:
(MergeQueueFactoryBase.__init__): Add ValidateRemote.
* Tools/CISupport/ews-build/factories_unittest.py:
(TestExpectedBuildSteps):
* Tools/CISupport/ews-build/steps.py:
(ValidateRemote.__init__): Set default summary.
(ValidateRemote.start): Check if the branch as owned by the target remote exists
on the branch as owned by origin.
(ValidateRemote.getResultSummary): Set the step summary.
(ValidateRemote.evaluateCommand): Invert return code. If the branch as owned by the
target remote exists on origin, fail the build and block the PR.
(ValidateRemote.doStepIf): Skip step for patches and origin-targeting PRs.
(ValidateRemote.hideStepIf): Hide step if it's being skipped.
* Tools/CISupport/ews-build/steps_unittest.py:

Canonical link: https://commits.webkit.org/252584@main
